pkg/controllers/provisioning/scheduling/machine.go

```diff
@@ -129,10 +129,122 @@ func InstanceTypeList(instanceTypeOptions []*cloudprovider.InstanceType) string
-	})
+type filterResults struct {
+	remaining []*cloudprovider.InstanceType
+	// Each of these three flags indicates if that particular criteria was met by at least one instance type
+	requirementsMet bool
+	fits            bool
+	hasOffering     bool
+
+	// requirementsAndFits indicates if a single instance type met the scheduling requirements and had enough resources
+	requirementsAndFits bool
+	// requirementsAndOffering indicates if a single instance type met the scheduling requirements and was a required offering
+	requirementsAndOffering bool
+	// fitsAndOffering indicates if a single instance type had enough resources and was a required offering
+	fitsAndOffering bool
+
+	requests v1.ResourceList
+}
+
+// FailureReason returns a presentable string explaining why all instance types were filtered out
+//
+//nolint:gocyclo
+func (r filterResults) FailureReason() string {
+	if len(r.remaining) > 0 {
+		return ""
+	}
+
+	// no instance type met any of the three criteria, meaning each criteria was enough to completely prevent
+	// this pod from scheduling
+	if !r.requirementsMet && !r.fits && !r.hasOffering {
+		return "no instance type met the scheduling requirements or had enough resources or had a required offering"
+	}
+
+	// check the other pairwise criteria
+	if !r.requirementsMet && !r.fits {
+		return "no instance type met the scheduling requirements or had enough resources"
+	}
+
+	if !r.requirementsMet && !r.hasOffering {
+		return "no instance type met the scheduling requirements or had a required offering"
+	}
+
+	if !r.fits && !r.hasOffering {
+		return "no instance type had enough resources or had a required offering"
+	}
+
+	// and then each individual criteria. These are sort of the same as above in that each one indicates that no
+	// instance type matched that criteria at all, so it was enough to exclude all instance types.  I think it's
+	// helpful to have these separate, since we can report the multiple excluding criteria above.
+	if !r.requirementsMet {
+		return "no instance type met all requirements"
+	}
+
+	if !r.fits {
+		msg := "no instance type has enough resources"
+		// special case for a user typo I saw reported once
+		if r.requests.Cpu().Cmp(resource.MustParse("1M")) >= 0 {
+			msg += " (CPU request >= 1 Million, m vs M typo?)"
+		}
+		return msg
+	}
+
+	if !r.hasOffering {
+		return "no instance type has the required offering"
+	}
+
+	// see if any pair of criteria was enough to exclude all instances
+	if r.requirementsAndFits {
+		return "no instance type which met the scheduling requirements and had enough resources, had a required offering"
+	}
+	if r.fitsAndOffering {
+		return "no instance type which had enough resources and the required offering met the scheduling requirements"
+	}
+	if r.requirementsAndOffering {
+		return "no instance type which met the scheduling requirements and the required offering had the required resources"
+	}
```

Comment:
I think this evolution of log messaging is definitely an improvement, and gives more visibility into what decisions what karpenter + k8s scheduler is doing.

Would it be more readable if we simply returned a more transparent log message w/ the three dimensions and their true/false values? I like the human-centric structure of a sentence, but formulating the combinatoric possibilities of failures here as discrete sentences is hard to get right from a clarity of language point of view.